SemanticVersion now handles metadata with no pre-release label present

PDQDakota opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

This brings the class up to par with the standard. This change uses the official RegEx that the semver project has for validating versions.

The test expecting `1.0.0+-` to fail was removed as this is a valid semver 2.0 version string as per the semver 2.0 RegEx.

https://regex101.com/r/s3wbbH/4/

<!-- Summarize your PR between here and the checklist. -->

## PR Context

This is to resolve #14605 which is an issue that affects me.

<!-- Provide a little reasoning as to why this Pull Request helps and why you have opened it. -->

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed: <!-- Number/link of that issue here -->

PDQDakota committed:
SemanticVersion now handles metadata with no pre-release label present

This brings the class up to par with the standard. This change uses the
official RegEx that the semver project has for validating versions.

The test expecting 1.0.0+- to fail was removed as this is a valid
semver 2.0 version string as per the semver 2.0 RegEx.

iSazonov commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -467,19 +471,42 @@ public SemanticVersion(string version)
         {
-            if (!string.IsNullOrEmpty(preReleaseLabel))
+            if (string.IsNullOrEmpty(preReleaseLabel) && string.IsNullOrEmpty(buildLabel))
             {
-                if (!Regex.IsMatch(preReleaseLabel, LabelUnitRegEx)) throw new FormatException(nameof(preReleaseLabel));
-
-                PreReleaseLabel = preReleaseLabel;
+                return;
```
Since we already check three first parameters in base constructor I think it is better follow this and check preReleaseLabel and buildLabel separately too. We have no need to complicate.

If you see a need to be very strong here we should use the same approach in all constructors - build a string (make sense to use ToString() method) and check the string with IsValid() method.

PDQDakota replied:
Makes sense to me, I'll make those changes and push when ready.

PDQDakota committed:
Simplified SemanticVersion Constructor Impl

This removes the check of the version numbers that the base constructor
already handled. Now it only checks that the preReleaseLabel and
buildLabel parameters are valid for a SemVer 2 string.

PDQDakota committed:
Forgot to remove the 'P's

The C# regex engine does not use 'P's to designate capture groups.
Removed them so PowerShell can build.

PDQDakota committed:
Simplification is working now

I made a mistake and didn't build Pwsh before running the Pester tests
last time. I've now fixed the issues and all tests are passing.

anmenaga commented:
Restarted Linux CIs that failed in unrelated PSGallery cert issue.

anmenaga commented:
@daxian-dbw @SteveL-MSFT @TravisEz13 Can you please review this? I recall we previously had some difficulties around semantic versioning, so I want to make sure this change won't have unwanted side effects. Thank you.

PDQDakota committed:
Added More SemVer Tests

This adds tests for non-ASCII and non-alphanumeric characters in the
build metadata and pre-release labels.

PDQDakota commented on test/powershell/engine/Basic/SemanticVersion.Tests.ps1:
```diff
@@ -188,7 +196,6 @@ Describe "SemanticVersion api tests" -Tags 'CI' {
             @{ name = "Missing parts";               errorId = "FormatException";    expectedResult = $false; version = "1.0.+alpha" }
             @{ name = "Missing parts";               errorId = "FormatException";    expectedResult = $false; version = "1.0.0-alpha+" }
             @{ name = "Missing parts";               errorId = "FormatException";    expectedResult = $false; version = "1.0.0-+" }
-            @{ name = "Missing parts";               errorId = "FormatException";    expectedResult = $false; version = "1.0.0+-" }
```
It's been added.

SteveL-MSFT commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -469,16 +475,19 @@ public SemanticVersion(int major, int minor, int patch, string preReleaseLabel,
         {
             if (!string.IsNullOrEmpty(preReleaseLabel))
             {
-                if (!Regex.IsMatch(preReleaseLabel, LabelUnitRegEx)) throw new FormatException(nameof(preReleaseLabel));
```
It looks like the old API would throw if the labels don't meet the semver requirement, but this new change will simply ignore them.  Was it a deliberate decision to not throw?

iSazonov replied:
Oops, my comment above was not addressed :-(

SteveL-MSFT commented on src/System.Management.Automation/engine/PSVersionInfo.cs:
```diff
@@ -501,11 +510,12 @@ public SemanticVersion(int major, int minor, int patch, string label)
             // 2) 'label' starts with letter or digit.
```
Can you fix the typo above?  `SymVer`

TravisEz13 committed:
change SymVer to SemVer
